doc/reference/trd-storage-permissions.md

```diff
@@ -0,0 +1,183 @@
+
+
+1 Introduction
+-------------------------------
+
+Tock applications need to be able to store persistent state. Additionally,
+applications need to be able to keep data private from other applications. The
+kernel should also be able to allow specific applications to read and modify
+state from other applications.
+
+This requires a method for assigning applications persistent identifiers,
+a mechanism for granting storage permissions to specific applications,
+and kernel abstractions for implementing storage capsules that respect
+the storage permissions.
+
+
+2 Scope
+-------------------------------
+
+This document only describes the permission architecture in Tock for supporting
+application persistent storage. This document does not include specific types of
+persistent storage (e.g., flash, FRAM, etc.) or storage access abstractions
+(e.g., block-access, byte-access, etc.).
+
+
+3 Stored State Identifiers
+-------------------------------
+
+All shared persistent storage implementations must store a 32 bit identifier
+with each stored object to mark the application that created the stored object.
+
+When applications write data, their ShortId must be used as the identifier. When
+the kernel writes data, the identifier must be 0.
+
+
+4 Permissions
+-------------------------------
+
+All persistent application data is labeled based on the application which wrote
+the data.
+Applications can read and modify data with suitable
+permissions.
+
+There are three types of permissions:
+
+1. **Write**: The application can write data.
+1. **Read**: The application can read data.
+1. **Modify**: The application can modify existing data.
+
+Each permission type is independent. For example, an application can be given
+read permission for specific data but not be able to write new data itself.
+
+Write is a boolean permission. An application either has permission to write or
+it does not.
+
+Read and Modify permissions are tuples of `(the permission type, stored state
+identifier)`. These permissions only exist as associated with a particular
+stored state identifier. That is, a Read permission gives an application
+permission to read only stored state marked with the associated stored state
+identifier, and a Modify permission gives an application permission to modify
+only stored state marked with the associated stored state identifier.
+
+
+5 Requirements
+-------------------------------
+
+The Tock storage model imposes the following requirements:
+
+1. Applications are given separate write, read, and modify permissions.
+2. The label stored with the persistent data when the data are written is the
+   application's short AppID.
+3. Applications without a `ShortId::Fixed` cannot access (i.e.,
+   read/write/modify) any persistent storage.
+4. How permissions are mapped to applications must be customizable for different
+   Tock kernels.
+
+Additionally, the kernel itself can be given permission to store state.
```

Comment:
What's the default here? Is kernel storage opt-in or opt-out?

Can capsules rely on kernel storage?

Reply:
> What's the default here? Is kernel storage opt-in or opt-out?

There is no default. It is up to the kernel.

> Can capsules rely on kernel storage?

Sure. In the same way that a capsule can rely on, say, a RNG.

Reply:
Okay, I think what my brain was circling is something like, "the core kernel MUST NOT rely on the presence of kernel storage for correct operation, kernel storage may only be used by capsules optionally included by board authors"

Reply:
I'm not sure how this relates to storage permissions. The permissions part is we have a way to differentiate the kernel from apps. But how the kernel is implemented in general seems out of scope.